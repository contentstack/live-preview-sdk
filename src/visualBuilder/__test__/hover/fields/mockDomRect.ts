export const mockDomRect = {
    singleHorizontal: () => ({
        bottom: 88.3984375,
        height: 54.3984375,
        left: 34,
        right: 862,
        top: 34,
        width: 828,
        x: 34,
        y: 34,
    }),
    singleLeft: () => ({
        x: 51,
        y: 51,
        width: 27.7734375,
        height: 20.3984375,
        top: 51,
        bottom: 71.3984375,
        left: 51,
        right: 78.7734375,
    }),
    singleRight: () => ({
        x: 110.7734375,
        y: 51,
        width: 53.59375,
        height: 20.3984375,
        top: 51,
        bottom: 71.3984375,
        left: 110.7734375,
        right: 164.3671875,
    }),
    singleVertical: () => ({
        bottom: 195.1953125,
        height: 90.796875,
        left: 34,
        right: 862,
        top: 104.3984375,
        width: 828,
        x: 34,
        y: 104.3984375,
    }),
    singleTop: () => ({
        x: 51,
        y: 121.3984375,
        width: 794,
        height: 20.3984375,
        top: 121.3984375,
        bottom: 141.796875,
        left: 51,
        right: 845,
    }),
    singleBottom: () => ({
        x: 51,
        y: 157.796875,
        width: 794,
        height: 20.3984375,
        top: 157.796875,
        bottom: 178.1953125,
        left: 51,
        right: 845,
    }),
    multipleLeft: () => ({
        x: 51,
        y: 228.1953125,
        width: 87.59375,
        height: 90.796875,
        top: 228.1953125,
        bottom: 318.9921875,
        left: 51,
        right: 138.59375,
    }),
    multipleRight: () => ({
        x: 170.59375,
        y: 228.1953125,
        width: 87.59375,
        height: 90.796875,
        top: 228.1953125,
        bottom: 318.9921875,
        left: 170.59375,
        right: 258.1875,
    }),
    multipleTop: () => ({
        x: 51,
        y: 368.9921875,
        width: 794,
        height: 90.796875,
        top: 368.9921875,
        bottom: 459.7890625,
        left: 51,
        right: 845,
    }),
    multipleBottom: () => ({
        x: 51,
        y: 475.7890625,
        width: 794,
        height: 90.796875,
        top: 475.7890625,
        bottom: 566.5859375,
        left: 51,
        right: 845,
    }),
    multipleChildTop: () => ({
        x: 68,
        y: 385.9921875,
        width: 760,
        height: 20.3984375,
        top: 385.9921875,
        bottom: 406.390625,
        left: 68,
        right: 828,
    }),
    multipleChildBottom: () => ({
        x: 68,
        y: 422.390625,
        width: 760,
        height: 20.3984375,
        top: 422.390625,
        bottom: 442.7890625,
        left: 68,
        right: 828,
    }),
};
