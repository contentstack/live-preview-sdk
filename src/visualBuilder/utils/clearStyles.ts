const originalStyles = new WeakMap<
    HTMLElement,
    {
        visibility?: string;
        transition?: string;
        animation?: string;
    }
>();

export function clearVisibelityStyles(element: HTMLElement) {
    const originalStyleValues = {
        visibility: element.style.visibility,
        transition: element.style.transition,
        animation: element.style.animation,
    };

    originalStyles.set(element, originalStyleValues);

    element.style.visibility = "hidden";
    element.style.transition = "none";
    element.style.animation = "none";
}

export function restoreVisibelityStyles(element: HTMLElement) {
    const storedStyles = originalStyles.get(element);

    if (storedStyles) {
        element.style.visibility = storedStyles.visibility || "";
        element.style.transition = storedStyles.transition || "";
        element.style.animation = storedStyles.animation || "";

        originalStyles.delete(element);
    }
}
