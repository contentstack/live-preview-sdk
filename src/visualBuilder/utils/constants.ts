import { FieldDataType } from "./types/index.types";

export const numericInputRegex = /^-?\d*(\.\d*)?([eE][-+]?\d*)?$/;

export const VISUAL_BUILDER_FIELD_TYPE_ATTRIBUTE_KEY = "data-cslp-field-type";

export const VISUAL_BUILDER_CHANNEL_ID = "visual-builder";

export const LIVE_PREVIEW_OUTLINE_WIDTH_IN_PX = 2;

// Minimum distance from top edge to prevent toolbar from being hidden
export const TOP_EDGE_BUFFER = 42;
export const RIGHT_EDGE_BUFFER = 180;

export const TOOLBAR_EDGE_BUFFER = 8;

export const DATA_CSLP_ATTR_SELECTOR = "data-cslp";

/**
 * The field that can be directly modified using contenteditable=true.
 * This includes all text fields like title and numbers.
 */
export const ALLOWED_INLINE_EDITABLE_FIELD: FieldDataType[] = [
    FieldDataType.SINGLELINE,
    FieldDataType.MULTILINE,
    FieldDataType.NUMBER,
];

export const ALLOWED_MODAL_EDITABLE_FIELD: FieldDataType[] = [
    FieldDataType.HTML_RTE,
    FieldDataType.MARKDOWN_RTE,
    FieldDataType.JSON_RTE,
    FieldDataType.CUSTOM_FIELD,
    FieldDataType.LINK,
    FieldDataType.ISODATE,
    FieldDataType.URL,
];

export const ALLOWED_REPLACE_FIELDS: FieldDataType[] = [
    FieldDataType.REFERENCE,
    FieldDataType.FILE,
];

export const DEFAULT_MULTIPLE_FIELDS: FieldDataType[] = [
    FieldDataType.GLOBAL_FIELD,
    FieldDataType.GROUP,
    FieldDataType.BLOCK,
];

export const mentionLimit = 20;
export const maxMessageLength = 500;
