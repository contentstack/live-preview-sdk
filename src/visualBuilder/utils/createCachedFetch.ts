/**
 * Creates a cached async fetch function with support for any number of arguments
 * @param fetchFn - The async function to cache
 * @param uidResolver - Function that generates a unique ID from the arguments passed to fetchFn
 * @returns A cached version of the fetch function with the same signature
 */
export function createCachedFetch<TArgs extends any[], TResult>(
    fetchFn: (...args: TArgs) => Promise<TResult>,
    uidResolver: (...args: TArgs) => string = (...args) => JSON.stringify(args)
): {
    (...args: TArgs): Promise<TResult>;
    clearCache: () => void;
} {
    // Cache storage
    const cache = new Map<string, TResult>();
    // Track in-flight requests
    const pendingPromises = new Map<string, Promise<TResult>>();

    /**
     * The cached fetch function
     * @param args - Arguments to pass to the original fetch function
     * @returns Promise that resolves with the data
     */
    async function cachedFetch(...args: TArgs): Promise<TResult> {
        // Generate unique ID for these arguments
        const uid = uidResolver(...args);

        // Return cached value if available
        if (cache.has(uid)) {
            return cache.get(uid)!;
        }

        // Return existing promise if request is already in progress
        if (pendingPromises.has(uid)) {
            return pendingPromises.get(uid)!;
        }

        // Create new promise for this request
        const promise = fetchFn(...args)
            .then((data) => {
                // Store result in cache
                cache.set(uid, data);
                // Remove from pending
                pendingPromises.delete(uid);
                return data;
            })
            .catch((error) => {
                // Clean up on error
                pendingPromises.delete(uid);
                throw error;
            });

        // Store the promise
        pendingPromises.set(uid, promise);
        return promise;
    }

    // Add clearCache method to the function
    cachedFetch.clearCache = () => {
        cache.clear();
        pendingPromises.clear();
    };

    return cachedFetch;
}
