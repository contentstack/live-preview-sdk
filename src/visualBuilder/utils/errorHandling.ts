
export function hasPostMessageError(obj: any): boolean {
    return obj?.error === true;
}