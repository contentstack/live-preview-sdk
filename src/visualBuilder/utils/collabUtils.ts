import {
    ICommentState,
    IMentionedList,
    IMentionItem,
    IMessageDTO,
    IUserDTO,
    IUserState,
} from "../types/collab.types";
import { maxMessageLength, mentionLimit } from "./constants";
import { uniqBy } from "lodash-es";
import DOMPurify from "dompurify";

const escapeRegExp = (string: string): string => {
    return string.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
};

/**
 * Generates the title for the thread based on the number of comments.
 * @param {number} commentCount - The number of comments.
 * @returns {string} The title for the thread.
 */
export const getThreadTitle = (commentCount: number): string => {
    if (commentCount === 0) return "Add New Feedback";
    return commentCount === 1 ? "1 Feedback" : `${commentCount} Feedbacks`;
};

/**
 * returns the available email.
 * @param {IUserDTO} user - The user object.
 * @returns {string} The user's email.
 */
export const getUserName = (user: IUserDTO): string => {
    return user.email;
};

/**
 * Validates the comment length and the number of mentions.
 * @param {string} comment - The comment message.
 * @param {IMentionedList} toUsers - The list of mentioned users.
 * @returns {string} The error message if validation fails, otherwise an empty string.
 */
export const validateCommentAndMentions = (
    comment: string,
    toUsers: IMentionedList
): string => {
    if (comment.length > maxMessageLength) {
        return `Limit exceeded. You can have a maximum length of ${maxMessageLength} characters.`;
    }
    if (toUsers.length > mentionLimit) {
        return `Limit exceeded. You can tag a maximum of ${mentionLimit} users.`;
    }
    return "";
};

/**
 * Removes mentions that no longer exist in the message.
 * @param {string} message - The comment message.
 * @param {IMentionedList} toUsers - The list of mentioned users.
 * @returns {Object} The updated lists of mentioned users.
 */
export const filterOutInvalidMentions = (
    message: string,
    toUsers: IMentionedList
) => {
    const to_users_temp = toUsers.filter((user) =>
        message.includes(user.display)
    );

    return {
        toUsers: uniqBy(to_users_temp, "id"),
    };
};

/**
 * Replaces mention placeholders with display names in the comment message.
 * @param {IMessageDTO | undefined} comment - The comment object.
 * @param {IUserState} userState - The user state containing user and role maps.
 * @param {"text" | "html"} profile - The format for the output message, either plain text or HTML.
 * @returns {string | undefined} The formatted message or undefined if the comment is not provided.
 */
export const getMessageWithDisplayName = (
    comment: IMessageDTO | undefined | null,
    userState: IUserState,
    profile: "text" | "html"
): string | undefined => {
    if (!comment) return undefined;

    let tempText = comment.message;

    comment?.toUsers?.forEach((user) => {
        const userPattern = new RegExp(`{{${user}}}`, "g");
        const userData = userState.userMap[user];
        const replacement =
            profile === "html"
                ? `<b>@${userData.display || getUserName(userData)}</b>`
                : `@${userData.display || getUserName(userData)}`;
        tempText = tempText.replace(userPattern, replacement);
    });

    return tempText;
};

/**
 * Sanitizes HTML content to prevent XSS attacks.
 * @param {any} dirty - The unsanitized HTML content.
 * @returns {string} The sanitized HTML content.
 */
export const sanitizeData = (dirty: any): string => {
    return DOMPurify.sanitize(dirty, { USE_PROFILES: { html: true } });
};

/**
 * Constructs the comment body with mentions replaced by their unique identifiers.
 * @param {ICommentState} state - The state containing the comment and mentions.
 * @returns {Object} The comment body containing the sanitized message and mentioned users.
 */
export const getCommentBody = (state: ICommentState): ICommentState => {
    let finalMessage = state.message.trim().replace(/\s+/g, " ");
    const comment = {
        message: finalMessage,
        toUsers: [],
        images: [],
        createdBy: state.createdBy,
        author: state.author,
    };

    const updateMentionToUID = (
        entity: IMentionItem,
        result: Array<string>
    ) => {
        const displayName = entity.display;

        const escapedDisplayName = escapeRegExp(displayName);
        const regexUser = new RegExp(escapedDisplayName, "g");
        finalMessage = finalMessage.replace(regexUser, `{{${entity.id}}}`);
        result.push(entity.display);
    };

    state.toUsers?.forEach((user) => updateMentionToUID(user, comment.toUsers));

    comment.message = finalMessage;
    return comment;
};
