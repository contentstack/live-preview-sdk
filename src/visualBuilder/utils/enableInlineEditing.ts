import { throttle } from "lodash-es";
import { VisualBuilder } from "../index";
import {
    isEllipsisActive,
    generatePseudoEditableElement,
} from "../generators/generatePseudoEditableField";
import { VISUAL_BUILDER_FIELD_TYPE_ATTRIBUTE_KEY } from "./constants";
import { getMultilinePlaintext } from "./getMultilinePlaintext";
import { handleFieldInput, handleFieldKeyDown } from "./handleFieldMouseDown";
import { FieldDataType, VisualBuilderEditContext } from "./types/index.types";
import { updateFocussedState } from "./updateFocussedState";
import { pasteAsPlainText } from "./pasteAsPlainText";

export function enableInlineEditing({
    expectedFieldData,
    editableElement,
    fieldType,
    elements,
}: {
    expectedFieldData: any;
    editableElement: HTMLElement;
    fieldType: FieldDataType;
    elements: VisualBuilderEditContext;
}) {
    const { visualBuilderContainer, resizeObserver } = elements;
    let actualEditableField = editableElement as HTMLElement;

    VisualBuilder.VisualBuilderGlobalState.value.focusFieldValue =
        actualEditableField?.innerText;

    const elementComputedDisplay =
        window.getComputedStyle(actualEditableField).display;

    let textContent =
        (editableElement as HTMLElement).innerText ||
        editableElement.textContent ||
        "";

    if (fieldType === FieldDataType.MULTILINE) {
        textContent = getMultilinePlaintext(actualEditableField);
        actualEditableField.addEventListener("paste", pasteAsPlainText);
    }
    const expectedTextContent = expectedFieldData;
    if (
        (expectedTextContent && textContent !== expectedTextContent) ||
        isEllipsisActive(editableElement as HTMLElement)
    ) {
        // TODO: Testing will be done in the E2E.
        const pseudoEditableField = generatePseudoEditableElement(
            { editableElement: editableElement as HTMLElement },
            { textContent: expectedFieldData }
        );

        (editableElement as HTMLElement).style.visibility = "hidden";

        // set field type attribute to the pseudo editable field
        // ensures proper keydown handling similar to the actual editable field
        pseudoEditableField.setAttribute(
            VISUAL_BUILDER_FIELD_TYPE_ATTRIBUTE_KEY,
            fieldType
        );
        visualBuilderContainer.appendChild(pseudoEditableField);
        actualEditableField = pseudoEditableField;

        if (fieldType === FieldDataType.MULTILINE)
            actualEditableField.addEventListener("paste", pasteAsPlainText);

        // we will unobserve this in hideOverlay
        elements.resizeObserver.observe(pseudoEditableField);
    } else if (elementComputedDisplay === "inline") {
        // if the editable field is inline
        const onInlineElementInput = throttle(() => {
            const overlayWrapper = visualBuilderContainer.querySelector(
                ".visual-builder__overlay__wrapper"
            ) as HTMLDivElement;
            const focusedToolbar = visualBuilderContainer.querySelector(
                ".visual-builder__focused-toolbar"
            ) as HTMLDivElement;
            updateFocussedState({
                editableElement: actualEditableField,
                visualBuilderContainer,
                overlayWrapper,
                focusedToolbar,
                resizeObserver,
            });
        }, 200);
        actualEditableField.addEventListener("input", onInlineElementInput);
    }

    actualEditableField.setAttribute("contenteditable", "true");
    actualEditableField.addEventListener("input", handleFieldInput);
    actualEditableField.addEventListener("keydown", handleFieldKeyDown);
    // focus on the contenteditable element to start accepting input
    actualEditableField.focus();

    return;
}
