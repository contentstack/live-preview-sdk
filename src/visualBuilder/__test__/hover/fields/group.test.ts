import { getFieldSchemaMap } from "../../../../__test__/data/fieldSchemaMap";
import { sleep, waitForHoverOutline } from "../../../../__test__/utils";
import Config from "../../../../configManager/configManager";
import { FieldSchemaMap } from "../../../utils/fieldSchemaMap";
import { mockDomRect } from "./mockDomRect";
import { VisualBuilder } from "../../../index";
import { screen } from "@testing-library/preact";
import { act } from "@testing-library/preact";

vi.mock("../../../utils/visualBuilderPostMessage", async () => {
    const { getAllContentTypes } = await vi.importActual<
        typeof import("../../../../__test__/data/contentType")
    >("../../../../__test__/data/contentType");
    const contentTypes = getAllContentTypes();
    return {
        __esModule: true,
        default: {
            send: vi.fn().mockImplementation((eventName: string) => {
                if (eventName === "init")
                    return Promise.resolve({
                        contentTypes,
                    });
                return Promise.resolve();
            }),
        },
    };
});

global.ResizeObserver = vi.fn().mockImplementation(() => ({
    observe: vi.fn(),
    unobserve: vi.fn(),
    disconnect: vi.fn(),
}));

describe("When an element is hovered in visual builder mode", () => {
    let mousemoveEvent: Event;

    beforeAll(() => {
        FieldSchemaMap.setFieldSchema(
            "all_fields",
            getFieldSchemaMap().all_fields
        );
    });

    beforeEach(() => {
        Config.reset();
        Config.set("mode", 2);
        mousemoveEvent = new Event("mousemove", {
            bubbles: true,
            cancelable: true,
        });
    });

    afterEach(() => {
        vi.clearAllMocks();
        document.getElementsByTagName("html")[0].innerHTML = "";
    });

    afterAll(() => {
        Config.reset();
    });

    describe("group field", () => {
        let groupField: HTMLDivElement;
        let nestedSingleLine: HTMLParagraphElement;
        let visualBuilder: VisualBuilder;

        beforeEach(() => {
            groupField = document.createElement("div");
            groupField.setAttribute(
                "data-cslp",
                "all_fields.bltapikey.en-us.group"
            );

            groupField.getBoundingClientRect = vi
                .fn()
                .mockReturnValue(mockDomRect.singleLeft());

            nestedSingleLine = document.createElement("p");
            nestedSingleLine.setAttribute(
                "data-cslp",
                "all_fields.bltapikey.en-us.group.single_line"
            );

            nestedSingleLine.getBoundingClientRect = vi
                .fn()
                .mockReturnValue(mockDomRect.singleLeft());

            groupField.appendChild(nestedSingleLine);
            document.body.appendChild(groupField);

            visualBuilder = new VisualBuilder();
        });

        afterEach(() => {
            visualBuilder.destroy();
        });

        test("should have outline and custom cursor", async () => {
            await act(async () => {
                groupField.dispatchEvent(mousemoveEvent);
            });
            await waitForHoverOutline();
            const hoverOutline = document.querySelector(
                "[data-testid='visual-builder__hover-outline']"
            );
            expect(hoverOutline).toHaveAttribute('style');

            const customCursor = document.querySelector(
                `[data-testid="visual-builder__cursor"]`
            );

            expect(customCursor).toHaveAttribute('data-icon', 'group');
            expect(customCursor?.classList.contains("visible")).toBeTruthy();
        });

        test("should have a outline and custom cursor on the nested single line", async () => {
            const singleLine = document.createElement("p");
            singleLine.setAttribute(
                "data-cslp",
                "all_fields.bltapikey.en-us.group.single_line"
            );

            singleLine.getBoundingClientRect = vi
                .fn()
                .mockReturnValue(mockDomRect.singleLeft());

            groupField.appendChild(singleLine);

            await act(async () => {
                singleLine.dispatchEvent(mousemoveEvent);
            });
            await waitForHoverOutline();
            const hoverOutline = document.querySelector(
                "[data-testid='visual-builder__hover-outline']"
            );
            expect(hoverOutline).toHaveAttribute('style');

            const customCursor = document.querySelector(
                `[data-testid="visual-builder__cursor"]`
            );

            expect(customCursor).toHaveAttribute('data-icon', 'singleline');
            expect(customCursor?.classList.contains("visible")).toBeTruthy();
        });
    });

    describe("group field (multiple)", () => {
        let container: HTMLDivElement;
        let firstGroupField: HTMLDivElement;
        let firstNestedMultiLine: HTMLParagraphElement;
        let secondGroupField: HTMLDivElement;
        let visualBuilder: VisualBuilder;

        beforeEach(() => {
            container = document.createElement("div");
            container.setAttribute(
                "data-cslp",
                "all_fields.bltapikey.en-us.group_multiple_"
            );

            container.getBoundingClientRect = vi
                .fn()
                .mockReturnValue(mockDomRect.singleHorizontal());

            firstGroupField = document.createElement("div");
            firstGroupField.setAttribute(
                "data-cslp",
                "all_fields.bltapikey.en-us.group_multiple_.0"
            );

            firstGroupField.getBoundingClientRect = vi
                .fn()
                .mockReturnValue(mockDomRect.singleLeft());

            firstNestedMultiLine = document.createElement("p");
            firstNestedMultiLine.setAttribute(
                "data-cslp",
                "all_fields.bltapikey.en-us.group_multiple_.0.multi_line"
            );

            firstNestedMultiLine.getBoundingClientRect = vi
                .fn()
                .mockReturnValue(mockDomRect.singleLeft());

            secondGroupField = document.createElement("div");
            secondGroupField.setAttribute(
                "data-cslp",
                "all_fields.bltapikey.en-us.group_multiple_.1"
            );

            secondGroupField.getBoundingClientRect = vi
                .fn()
                .mockReturnValue(mockDomRect.singleRight());

            container.appendChild(firstGroupField);
            container.appendChild(secondGroupField);

            firstGroupField.appendChild(firstNestedMultiLine);

            document.body.appendChild(container);

            visualBuilder = new VisualBuilder();
        });

        afterEach(() => {
            visualBuilder.destroy();
        });

        test("should have outline and custom cursor", async () => {
            await act(async () => {
                container.dispatchEvent(mousemoveEvent);
            });
            await waitForHoverOutline();
            const hoverOutline = document.querySelector(
                "[data-testid='visual-builder__hover-outline']"
            );
            expect(hoverOutline).toHaveAttribute('style');

            const customCursor = document.querySelector(
                `[data-testid="visual-builder__cursor"]`
            );

            expect(customCursor).toHaveAttribute('data-icon', 'group');
            expect(customCursor?.classList.contains("visible")).toBeTruthy();

            await act(async () => {
                firstNestedMultiLine.dispatchEvent(mousemoveEvent);
            });
            await waitForHoverOutline();

            const newCustomCursor = document.querySelector(
                `[data-testid="visual-builder__cursor"]`
            );
            expect(newCustomCursor).toHaveAttribute('data-icon', 'multiline');
            expect(newCustomCursor?.classList.contains("visible")).toBeTruthy();
        });
    });
});
