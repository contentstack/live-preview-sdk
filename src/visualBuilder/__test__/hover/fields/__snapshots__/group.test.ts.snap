// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`When an element is hovered in visual builder mode > group field (multiple) > should have custom cursor 1`] = `
<div
  class="go1688149084 visual-builder__cursor visible"
  data-testid="visual-builder__cursor"
/>
`;

exports[`When an element is hovered in visual builder mode > group field (multiple) > should have outline on the nested field 1`] = `
<p
  data-cslp="all_fields.bltapikey.en-us.group_multiple_.0.multi_line"
/>
`;

exports[`When an element is hovered in visual builder mode > group field (multiple) > should have outline on the nested field 2`] = `
<div
  class="visual-builder__hover-outline visual-builder__hover-outline--unclickable go3177626044 go1719569523"
  data-testid="visual-builder__hover-outline"
  style="top: 51px; left: 51px; width: 27.7734375px; height: 20.3984375px;"
/>
`;

exports[`When an element is hovered in visual builder mode > group field > should have a outline on the nested single line 1`] = `
<p
  data-cslp="all_fields.bltapikey.en-us.group.single_line"
/>
`;

exports[`When an element is hovered in visual builder mode > group field > should have a outline on the nested single line 2`] = `
<div
  class="visual-builder__hover-outline visual-builder__hover-outline--unclickable go3177626044 go1719569523"
  data-testid="visual-builder__hover-outline"
  style="top: 51px; left: 51px; width: 27.7734375px; height: 20.3984375px;"
/>
`;

exports[`When an element is hovered in visual builder mode > group field > should have custom cursor 1`] = `
<div
  class="go1688149084 visual-builder__cursor visible"
  data-icon="group"
  data-testid="visual-builder__cursor"
>
  <div
    class="visual-builder__cursor-wrapper go4230334221"
  >
    <div
      class="visual-builder__cursor-pointer go3046085498"
    >
      <svg
        fill="none"
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0.0167948 0.602214C-0.0784404 0.246792 0.246794 -0.0784402 0.602216 0.0167948L15.6457 4.04767C16.0914 4.1671 16.1259 4.78625 15.6962 4.95447L8.17161 7.90048C8.04753 7.94906 7.94936 8.04723 7.90078 8.17131L4.95449 15.6962C4.78627 16.1259 4.16712 16.0914 4.04769 15.6457L0.0167948 0.602214Z"
          fill="#5D50BE"
        />
      </svg>
    </div>
    <div
      class="visual-builder__cursor-icon go4261356330"
    >
      <svg
        fill="none"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M2.8125 3.75C2.8125 3.23223 3.23223 2.8125 3.75 2.8125C4.26777 2.8125 4.6875 3.23223 4.6875 3.75C4.6875 4.26777 4.26777 4.6875 3.75 4.6875C3.23223 4.6875 2.8125 4.26777 2.8125 3.75ZM3.75 1.6875C4.69408 1.6875 5.49001 2.32181 5.73486 3.1875H18.2651C18.51 2.32181 19.3059 1.6875 20.25 1.6875C21.3891 1.6875 22.3125 2.61091 22.3125 3.75C22.3125 4.69408 21.6782 5.49002 20.8125 5.73486V18.2651C21.6782 18.51 22.3125 19.3059 22.3125 20.25C22.3125 21.3891 21.3891 22.3125 20.25 22.3125C19.3059 22.3125 18.51 21.6782 18.2651 20.8125H5.73486C5.49001 21.6782 4.69408 22.3125 3.75 22.3125C2.61091 22.3125 1.6875 21.3891 1.6875 20.25C1.6875 19.3059 2.32181 18.51 3.1875 18.2651V5.73486C2.32181 5.49001 1.6875 4.69408 1.6875 3.75C1.6875 2.61091 2.61091 1.6875 3.75 1.6875ZM20.25 4.6875C20.7678 4.6875 21.1875 4.26777 21.1875 3.75C21.1875 3.23223 20.7678 2.8125 20.25 2.8125C19.7322 2.8125 19.3125 3.23223 19.3125 3.75C19.3125 4.26777 19.7322 4.6875 20.25 4.6875ZM18.2651 4.3125C18.4594 4.99938 19.0006 5.54059 19.6875 5.73486V18.2651C19.0006 18.4594 18.4594 19.0006 18.2651 19.6875H5.73486C5.54059 19.0006 4.99938 18.4594 4.3125 18.2651V5.73486C4.99938 5.54059 5.54059 4.99938 5.73486 4.3125H18.2651ZM19.3125 20.25C19.3125 19.7322 19.7322 19.3125 20.25 19.3125C20.7678 19.3125 21.1875 19.7322 21.1875 20.25C21.1875 20.7678 20.7678 21.1875 20.25 21.1875C19.7322 21.1875 19.3125 20.7678 19.3125 20.25ZM4.6875 20.25C4.6875 19.7322 4.26777 19.3125 3.75 19.3125C3.23223 19.3125 2.8125 19.7322 2.8125 20.25C2.8125 20.7678 3.23223 21.1875 3.75 21.1875C4.26777 21.1875 4.6875 20.7678 4.6875 20.25ZM9.75 7.6875C9.75 6.96263 10.3376 6.375 11.0625 6.375H15.5625C16.2874 6.375 16.875 6.96263 16.875 7.6875V13.6875C16.875 14.4124 16.2874 15 15.5625 15H14.625V15.9375C14.625 16.6624 14.0374 17.25 13.3125 17.25H8.8125C8.08763 17.25 7.5 16.6624 7.5 15.9375V9.1875C7.5 8.46263 8.08763 7.875 8.8125 7.875H9.75V7.6875ZM9.75 9H8.8125C8.70895 9 8.625 9.08395 8.625 9.1875V15.9375C8.625 16.0411 8.70895 16.125 8.8125 16.125H13.3125C13.4161 16.125 13.5 16.0411 13.5 15.9375V15H11.0625C10.3376 15 9.75 14.4124 9.75 13.6875V9ZM11.0625 7.5C10.9589 7.5 10.875 7.58395 10.875 7.6875V13.6875C10.875 13.7911 10.9589 13.875 11.0625 13.875H15.5625C15.6661 13.875 15.75 13.7911 15.75 13.6875V7.6875C15.75 7.58395 15.6661 7.5 15.5625 7.5H11.0625Z"
          fill="white"
          fill-rule="evenodd"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`When an element is hovered in visual builder mode > group field > should have custom cursor on the nested single line 1`] = `
<div
  class="go1688149084 visual-builder__cursor visible"
  data-icon="singleline"
  data-testid="visual-builder__cursor"
>
  <div
    class="visual-builder__cursor-wrapper go4230334221"
  >
    <div
      class="visual-builder__cursor-pointer go3046085498"
    >
      <svg
        fill="none"
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0.0167948 0.602214C-0.0784404 0.246792 0.246794 -0.0784402 0.602216 0.0167948L15.6457 4.04767C16.0914 4.1671 16.1259 4.78625 15.6962 4.95447L8.17161 7.90048C8.04753 7.94906 7.94936 8.04723 7.90078 8.17131L4.95449 15.6962C4.78627 16.1259 4.16712 16.0914 4.04769 15.6457L0.0167948 0.602214Z"
          fill="#5D50BE"
        />
      </svg>
    </div>
    <div
      class="visual-builder__cursor-icon go4261356330"
    >
      <svg
        fill="none"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M5.94958 7.8294C5.85419 7.6199 5.64077 7.48967 5.41084 7.50064C5.1809 7.51162 4.98086 7.66158 4.90585 7.87921L3.19036 12.8566L2.2835 15.3717C2.17813 15.664 2.32962 15.9863 2.62186 16.0917C2.9141 16.197 3.23643 16.0455 3.34181 15.7533L3.95144 14.0625H7.53908L8.29768 15.7888C8.42266 16.0732 8.75454 16.2025 9.03895 16.0775C9.32336 15.9525 9.4526 15.6206 9.32762 15.3362L8.22073 12.8173L8.21769 12.8105L5.94958 7.8294ZM7.03937 12.9375L5.51002 9.57881L4.35242 12.9375H7.03937Z"
          fill="white"
          fill-rule="evenodd"
        />
        <path
          clip-rule="evenodd"
          d="M11.625 8.0625C11.625 7.75184 11.3732 7.5 11.0625 7.5C10.7518 7.5 10.5 7.75184 10.5 8.0625V15.5625C10.5 15.8732 10.7518 16.125 11.0625 16.125C11.3731 16.125 11.6249 15.8733 11.625 15.5627C12.0951 15.9158 12.6793 16.125 13.3125 16.125C14.8658 16.125 16.125 14.8658 16.125 13.3125C16.125 11.7592 14.8658 10.5 13.3125 10.5C12.6793 10.5 12.0951 10.7092 11.625 11.0623V8.0625ZM11.625 13.3116C11.625 13.3119 11.625 13.3122 11.625 13.3125C11.625 13.3128 11.625 13.3131 11.625 13.3134C11.6255 14.245 12.3808 15 13.3125 15C14.2445 15 15 14.2445 15 13.3125C15 12.3805 14.2445 11.625 13.3125 11.625C12.3808 11.625 11.6255 12.38 11.625 13.3116Z"
          fill="white"
          fill-rule="evenodd"
        />
        <path
          clip-rule="evenodd"
          d="M19.9448 10.5025C20.4222 10.4825 20.8969 10.5844 21.324 10.7988C21.6017 10.9381 21.7138 11.2762 21.5744 11.5538C21.4351 11.8315 21.0971 11.9436 20.8194 11.8043C20.5631 11.6757 20.2783 11.6145 19.9919 11.6265C19.7054 11.6385 19.4267 11.7233 19.1821 11.8729C18.9375 12.0224 18.735 12.2319 18.5938 12.4814C18.4526 12.731 18.3773 13.0124 18.375 13.2991C18.3728 13.5858 18.4436 13.8684 18.5808 14.1201C18.718 14.3719 18.9171 14.5845 19.1593 14.738C19.4015 14.8914 19.6788 14.9806 19.9651 14.9972C20.2513 15.0137 20.5371 14.9571 20.7953 14.8326C21.0752 14.6977 21.4114 14.8151 21.5463 15.095C21.6812 15.3748 21.5637 15.711 21.2839 15.846C20.8534 16.0535 20.3772 16.1479 19.9001 16.1203C19.423 16.0927 18.9609 15.944 18.5572 15.6883C18.1535 15.4325 17.8217 15.0781 17.593 14.6585C17.3643 14.2389 17.2463 13.768 17.2501 13.2901C17.2539 12.8123 17.3794 12.3433 17.6147 11.9274C17.8501 11.5115 18.1875 11.1624 18.5952 10.9131C19.0029 10.6638 19.4673 10.5225 19.9448 10.5025Z"
          fill="white"
          fill-rule="evenodd"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`When an element is hovered in visual builder mode > group field > should have outline 1`] = `
<div
  data-cslp="all_fields.bltapikey.en-us.group"
>
  <p
    data-cslp="all_fields.bltapikey.en-us.group.single_line"
  />
</div>
`;

exports[`When an element is hovered in visual builder mode > group field > should have outline 2`] = `
<div
  class="visual-builder__hover-outline visual-builder__hover-outline--unclickable go3177626044 go1719569523"
  data-testid="visual-builder__hover-outline"
  style="top: 51px; left: 51px; width: 27.7734375px; height: 20.3984375px;"
/>
`;
